routing/index_graph_starter.hpp

```diff
@@ -21,136 +26,180 @@ class IndexGraphStarter final
-  {
-    return m_graph.GetEstimator().CalcSegmentWeight(
-        segment, m_graph.GetRoadGeometry(segment.GetMwmId(), segment.GetFeatureId()));
-  }
+  double CalcSegmentWeight(Segment const & segment) const;
+  double CalcRouteSegmentWeight(std::vector<Segment> const & route, size_t segmentIndex) const;
 
-  bool IsLeap(NumMwmId mwmId) const
-  {
-    return mwmId != kFakeNumMwmId && mwmId != m_start.GetMwmId() && mwmId != m_finish.GetMwmId() &&
-           m_graph.GetEstimator().LeapIsAllowed(mwmId);
-  }
+  bool IsLeap(NumMwmId mwmId) const;
 
   static bool IsFakeSegment(Segment const & segment)
   {
     return segment.GetFeatureId() == kFakeFeatureId;
   }
 
 private:
-  void GetFakeToNormalEdges(FakeVertex const & fakeVertex, bool isOutgoing,
-                            vector<SegmentEdge> & edges);
-  void GetFakeToNormalEdge(FakeVertex const & fakeVertex, bool forward,
-                           vector<SegmentEdge> & edges);
-  void GetNormalToFakeEdge(Segment const & segment, FakeVertex const & fakeVertex,
-                           Segment const & fakeSegment, bool isOutgoing,
-                           vector<SegmentEdge> & edges);
+  static Segment GetFakeSegment(uint32_t i)
+  {
+    return Segment(kFakeNumMwmId, kFakeFeatureId, i, false);
+  }
+
+  class FakeVertex final
+  {
+  public:
+    enum class Type
+    {
+      PureFake,
+      Projection,
+      PartOfReal,
+    };
+    // For unit tests only.
+    FakeVertex(NumMwmId mwmId, uint32_t featureId, uint32_t segmentIdx, m2::PointD const & point)
+      : m_junctionFrom(point, feature::kDefaultAltitudeMeters)
+      , m_junctionTo(point, feature::kDefaultAltitudeMeters)
+    {
+    }
+
+    FakeVertex(Junction const & junctionFrom, Junction const & junctionTo, Type type)
+      : m_junctionFrom(junctionFrom), m_junctionTo(junctionTo), m_type(type)
+    {
+    }
+
+    FakeVertex(FakeVertex const &) = default;
+    FakeVertex() = default;
+
+    bool operator==(FakeVertex const & rhs) const
+    {
+      return m_junctionFrom == rhs.m_junctionFrom && m_junctionTo == rhs.m_junctionTo &&
+             m_type == rhs.m_type;
+    }
+
+    Type GetType() const { return m_type; }
+
+    Junction const & GetJunctionFrom() const { return m_junctionFrom; }
+    m2::PointD const & GetPointFrom() const { return m_junctionFrom.GetPoint(); }
+    Junction const & GetJunctionTo() const { return m_junctionTo; }
+    m2::PointD const & GetPointTo() const { return m_junctionTo.GetPoint(); }
+
+  private:
+    Junction m_junctionFrom;
+    Junction m_junctionTo;
+    Type m_type = Type::PureFake;
+  };
+
+  struct FakeGraph
+  {
+    std::map<Segment, std::set<Segment>> m_outgoing;
+    std::map<Segment, std::set<Segment>> m_ingoing;
+    std::map<Segment, FakeVertex> m_segmentToVertex;
+    std::map<Segment, Segment> m_fakeToReal;
+    std::map<Segment, std::set<Segment>> m_realToFake;
+  };
+
+  void AddFakeVertex(Segment const & existentSegment, Segment const & newSegment,
```

Comment:
Please, add a comment to the method. Three segments + a couple of bools require exhaustive documentation.

Reply:
done